distributed/scheduler.py

```diff
@@ -7488,36 +7532,92 @@ def decide_worker(
+            return ws
+
+        # Previous worker is fully assigned, so pick a new worker.
+        # Since this is a root-like task, we should ignore the placement of its dependencies while selecting workers.
+        # Every worker is going to end up running this type of task eventually, and any dependencies will have to be
+        # transferred to all workers, so there's no gain from only considering workers where the dependencies already live.
+        # Indeed, we _must_ consider all workers, otherwise we would keep picking the same "new" worker(s) every time,
+        # since there are only N workers to choose from that actually have the dependency (where N <= n_deps).
+        ignore_deps_while_picking = True
+        print(f"{ts.group_key} is root-like but {ws.name} is full - picking a new worker")
+
+    # Not a root-like task; pick the best worker among the valid workers
+    # that hold at least one dependency of this task.
     deps: set = ts._dependencies
+    dts: TaskState
     candidates: set
     assert all([dts._who_has for dts in deps])
-    if ts._actor:
-        candidates = set(all_workers)
+    if ignore_deps_while_picking:
+        candidates = valid_workers if valid_workers is not None else set(all_workers)
     else:
-        candidates = {wws for dts in deps for wws in dts._who_has}
-    if valid_workers is None:
-        if not candidates:
+        if ts._actor:
             candidates = set(all_workers)
-    else:
-        candidates &= valid_workers
-        if not candidates:
-            candidates = valid_workers
+        else:
+            candidates = {wws for dts in deps for wws in dts._who_has}
+        if valid_workers is None:
             if not candidates:
-                if ts._loose_restrictions:
-                    ws = decide_worker(ts, all_workers, None, objective)
-                return ws
+                candidates = set(all_workers)
+        else:
+            candidates &= valid_workers
+            if not candidates:
+                candidates = valid_workers
+                if not candidates:
+                    if ts._loose_restrictions:
+                        ws = decide_worker(ts, all_workers, None, objective, total_nthreads)
+                    return ws
 
     ncandidates: Py_ssize_t = len(candidates)
     if ncandidates == 0:
+        print(f"no candidates - {ts.group_key}")
         pass
     elif ncandidates == 1:
+        # NOTE: this is the ideal case: all the deps are already on the same worker.
+        # We did a good job in previous `decide_worker`s!
         for ws in candidates:
             break
+        print(f"1 candidate - {ts.group_key} -> {ws.name}")
     else:
         ws = min(candidates, key=objective)
+        print(f"picked worker - {ts.group_key} -> {ws.name}")
+
+    group._last_worker = ws
+    group._last_worker_tasks_left = math.floor(group_tasks_per_worker)
+    group._last_worker_priority = ts.priority
```

Comment:
Do we want to do this when we're not in a root-ish task situation?

Reply:
Yes, because we're also using this as our base case when `group._last_worker` is not set. So for the first root-ish task, we fall through to here, and then need to store our decision for the next `decide_worker` cycle. We could skip doing this if we've identified that the task is not root-like (which would save us from ever checking that again), though it might make the code a little more tangled.

Reply:
> So for the first root-ish task, we fall through to here

We can maybe guard this code block with a condition that we're root-ish then?

Reply:
Yes—though we currently skip determining if we're root-ish when `last_worker` is None. So we'd need some other sentinel value to distinguish "no last worker because we haven't checked yet if it's root-ish" from "no last worker because it's not root-ish".

Reply:
We currently check the following:

```python
        and group_tasks_per_worker > 1  # group is larger than cluster
        and (  # is a root-like task (task group is large, but depends on very few tasks)
            sum(map(len, group._dependencies)) < 5  # TODO what number
        )
```

I recommend that we define that as a variable at the top and then use it in a couple of places.  Would that work?